ddtrace/internal/datadog/profiling/dd_wrapper/src/interface.cpp

```diff
@@ -25,9 +25,27 @@ thread_local std::optional<Datadog::Sample> g_profile = std::nullopt;
 // When a fork is detected, we need to reinitialize this state.
 // This handler will be called in the single thread of the child process after the fork
 void
-ddup_fork_handler()
+ddup_postfork_child()
 {
     Datadog::Profile::mark_dirty();
+    Datadog::Uploader::unlock();
+}
+
+void
+ddup_postfork_parent()
+{
+    Datadog::Uploader::unlock();
+}
+
+// Since we don't control the internal state of libdatadog's exporter and we want to prevent state-tearing
+// after a `fork()`, we just outright prevent uploads from happening during a `fork()` operation.  Since a
+// new upload may be scheduled _just_ as `fork()` is entered, we also need to prevent new uploads from happening.
+// This mutex is released in both the child and the parent.
+void
+ddup_prefork()
+{
+    Datadog::Uploader::lock();
+    Datadog::Uploader::cancel_inflight();
```

Comment:
Is this synchronous or async?  If async, we might still have state tearing?

Reply:
Confirmed that this is sync, added comment indicating such